### ISSUE TITLE
Mix: JS API cannot comprehend solidity enums in function returns

### ISSUE THREAD

**Issue by Smithgift** [ethereum/cpp-ethereum#1801]
When a solidity function returns an enum, the JS API doesn't work, probably because that enum does not exist on the JS side.

Here's a representative selection from a terminal, calling an enum-returning function right out of the JS console in mix:
```
qml: {"jsonrpc":"2.0","method":"eth_sendTransaction","params":[{"to":"0x5b864aadd9ae77a178eef9728a24d28937dd5b77","data":"0x2399131c"}],"id":2}
/build/buildd/cpp-ethereum-0.9.13+780SNAPSHOT20150430134852trusty/mix/ContractCallDataEncoder.cpp(215): Throw in function QVariant dev::mix::ContractCallDataEncoder::decode(const dev::mix::SolidityType&, const bytes&)
Dynamic exception type: boost::exception_detail::clone_impl<dev::Exception>
std::exception::what: std::exception
[dev::tag_comment*] = Parameter declaration not found
qml: 
qml: http://localhost:47334/WebContainer.html:0: Uncaught SyntaxError: Unexpected end of input
```

**Comment by debris** [comment]
That's true, they are not supported in JS api. @chriseth can you guide me how enums should be encoded/decoded? It's not mentioned in the [docs](https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI)

**Comment by chriseth** [comment]
enums are not visible "to the outside". If you look at the json abi, you will see that they are automatically converted to suitable uint types.
This is the theory ;-)
Looks like a bug in the interface between mix and solidity. @arkpar, could you please take a look?

**Comment by Smithgift** [comment]
@chriseth Looking at the json abi straight from solc, a return type of an enum foo is actually `"enum foo"`

**Comment by chriseth** [comment]
This seems like a bug in `InterfaceHandler::getABIInterface`, it does not correctly convert the parameters to their external types.

**Comment by chriseth** [comment]
@Smithgift: Should be fixed on the solidity side. Would you be so kind and check whether it is fixed for you now all the way through the layers?

**Comment by Smithgift** [comment]
@chriseth: Mix is currently broken for me (c.f. #1874), but when it's working again I'll be sure to check again. Thank you all.

**Comment by Smithgift** [comment]
Update: Now that mix is a little more functional, it turns out that mix's end **is still** bugged. Trying to get a enum in a function return gives:

```
qml: {"jsonrpc":"2.0","method":"eth_sendTransaction","params":[{"to":"0x3a2a0a1f1d4a9bc70663cd7ec3b94db07d019857","data":"0x2399131c"}],"id":10}
/build/buildd/cpp-ethereum-0.9.20+841SNAPSHOT20150514134730trusty/mix/ContractCallDataEncoder.cpp(227): Throw in function QVariant dev::mix::ContractCallDataEncoder::decode(const dev::mix::SolidityType&, const bytes&)
Dynamic exception type: boost::exception_detail::clone_impl<dev::Exception>
std::exception::what: std::exception
[dev::tag_comment*] = Parameter declaration not found
qml: 
qml: qrc:///js/webthree.js:2784: Uncaught Error: Invalid JSON RPC response
```

**Comment by yann300** [comment]
@Smithgift should be fixed now by #2569.
